sharkfuser/include/fusili/asm_emitter.h

```diff
@@ -0,0 +1,420 @@
+// Copyright 2025 Advanced Micro Devices, Inc.
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===----------------------------------------------------------------------===//
+//
+// This file contains the inline definitions for all the MLIR assembly
+// generation methods on the `Graph`, `TensorAttr`, `INode` and derived node
+// classes. It is meant to be a common place for all things ASM emitter related
+// to make maintenance and future improvements easier.
+//
+// We use a combination of raw multi-line strings `R"(...)"` and `std::format`
+// (from c++20) to implement a simple templating system for generating MLIR
+// assembly code. This could be made better with a jinja2-like templating
+// system but for now this gets us mostly what we need.
+//
+// Caution: An important foot-gun with `std::format` is to forget to double the
+// brace for a literal `{` or `}`. i.e. always use `{{` for `{` and `}}` for `}`
+// to disambiguate from the `{}` that `std::format` uses for replacements.
+// If not you'll hit a compilation error like so:
+//    "error: call to consteval function 'std::basic_format_string<char, ...'"
+//    "is not a constant expression"
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef FUSILI_ASM_EMITTER_H
+#define FUSILI_ASM_EMITTER_H
+
+#include "fusili/attributes/tensor_attributes.h"
+#include "fusili/graph.h"
+#include "fusili/node/conv_node.h"
+#include "fusili/types.h"
+
+#include <cassert>
+#include <format>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace fusili {
+
+// Map from Fusili types to MLIR types.
+static const std::unordered_map<DataType, std::string> DataTypeToMlirTypeAsm = {
+    {DataType::Half, "f16"},       {DataType::BFloat16, "bf16"},
+    {DataType::Float, "f32"},      {DataType::Double, "f64"},
+    {DataType::Uint8, "ui8"},      {DataType::Int8, "si8"},
+    {DataType::Int16, "si16"},     {DataType::Int32, "si32"},
+    {DataType::Int64, "si64"},     {DataType::Boolean, "i1"},
+    {DataType::FP8E5M2, "f8E5M2"},
+};
+
+// Given a vector of ints, returns the MLIR assembly for the
+// `torch.constant.int` ops for each int value and the
+// `torch.prim.ListConstruct` op wrapping these into a single
+// value.
+//
+// For example if `getListOfIntOpsAsm` is called on these inputs:
+//    listOfInts: {1, 2}
+//    prefix: "stride"
+//    suffix: "conv"
+//
+// It generates the following MLIR assembly:
+//
+//   %stride_val_0_conv = torch.constant.int 1
+//   %stride_val_1_conv = torch.constant.int 2
+//   %stride_conv = torch.prim.ListConstruct
```

Comment:
We had discussed strategies for removing potential name conflicts, unless I'm missing something this doesn't include anything for that? Have you decided on a strategy?

Reply:
A few lines down there's a comment about that:
```
// The prefix is generally what attribute this refers to (e.g.
// padding, stride, dilation etc.) and the suffix is the node's
// unique name (for SSA disambiguation).
```
With this setup we basically only need to ensure the node attributes use unique names - by default this should be the case ([because](https://sourcegraph.com/github.com/nod-ai/shark-ai@d470b828374f121483a16589026cd263306f38fe/-/blob/sharkfuser/include/fusili/graph.h?L158)), but in the rare case when user inadvertently provides say two nodes with the same name, we will weed that out in the `Graph::validate()` call.